narwhal/primary/src/certificate_waiter.rs

```diff
@@ -3,226 +3,381 @@
-                    timer.as_mut().reset(Instant::now() + Duration::from_millis(GC_RESOLUTION));
-                    attempt_garbage_collection = true;
-                },
-
-                Ok(()) = self.rx_consensus_round_updates.changed() => {
-                    attempt_garbage_collection = true;
+                    debug!("Committee updated to {}", self.committee);
                 }
+            }
+        }
+    }
 
+    // Starts a task to fetch missing certificates from other primaries.
+    // A call to kick() can be triggered by a certificate with missing parents or the end of a fetch task.
+    // Each iterations of kick() updates the target rounds, and iterations will continue until there is no more target rounds to catch up to.
+    #[allow(clippy::mutable_key_type)]
+    fn kick(&mut self) {
+        let progression = match self.read_round_progression() {
+            Ok(progression) => progression,
+            Err(e) => {
+                warn!("Failed to read rounds per authority from the certificate store: {e}");
+                return;
             }
+        };
+        self.target.retain(|origin, target_round| {
+            let current_round = progression.get(origin).unwrap();
+            // Drop sync target when cert store already has an equal or higher round for the origin.
+            current_round < target_round
+        });
+        if self.target.is_empty() {
+            trace!("Certificates have caught up. Skip fetching.");
+            return;
+        }
 
-            // Either upon time-out or round change
-            if attempt_garbage_collection {
-                let round = *self.rx_consensus_round_updates.borrow();
-                if round > self.gc_depth {
-                    let gc_round = round - self.gc_depth;
-
-                    self.pending.retain(|_digest, (r, once_cancel)| {
-                        if *r <= gc_round {
-                            // note: this send can fail, harmlessly, if the certificate has been delivered (`notify_read`)
-                            // and the present code path fires before the corresponding `waiting` item is unpacked above.
-                            let _ = once_cancel
-                                .take()
-                                .expect("This should be protected by a write lock")
-                                .send(());
-                            false
-                        } else {
-                            true
-                        }
-                    });
+        let handle = self.create_fetch_task(progression);
+        let epoch = self.committee.epoch();
+        let metrics = self.metrics.clone();
+        let tx_kick_from_fetch = self.tx_kick_from_fetch.clone();
+
+        self.fetch_certificates_task = Some(tokio::task::spawn(async move {
+            metrics
+                .certificate_waiter_inflight_fetch
+                .with_label_values(&[&epoch.to_string()])
+                .set(1);
```

Comment:
insetad of set(1) and set(0) would it make sense to inc() and dec() here so you can detect if there are any issues with the code failing to limit this to one inflight fetch?